locktopus.js

```diff
@@ -0,0 +1,85 @@
+import Database from 'better-sqlite3'
+import lib from './lib/dmap/dmap.js'
+import { rpc }  from './rpc.js'
+import { existsSync, mkdirSync, readFileSync } from 'fs'
+
+let   config
+let   db
+let   last
+const lots = []
+const slot_idx = 0
+const save_idx = 1
+
+export const locktopus = {}
+
+locktopus.set_config = (new_config) => {
+    config = new_config
+}
+
+locktopus.init_db = (dir) => {
+    if (!existsSync(dir)) mkdirSync(dir)
+    db = new Database(`${dir}/locktopus.sqlite`, { verbose: console.log })
+    const create = db.prepare("CREATE TABLE IF NOT EXISTS locks(" +
+        "'when' INTEGER, 'slot' TEXT PRIMARY KEY, 'zone' TEXT, 'name' TEXT, 'meta' TEXT, 'data' TEXT )")
+    create.run();
+
+    lib.get = locktopus.monk_get.bind(null, db, lib.get)
+}
+
+locktopus.init_last = async () => {
+    const block = await rpc.getBlock(config.eth_rpc, 'latest')
+    last = block.timestamp
+}
+
+locktopus.look = async (path) => {
+    const dmap = await rpc.getFacade(config.eth_rpc, db)
+    const trace = await lib.walk2(dmap, path);
+    let [meta, data] = trace.slice(-1)[0]
+    if (lots.some(s => s[save_idx] === true)) await locktopus.save(trace, path)
+    console.log(`meta: ${meta}\ndata: ${data}`)
+}
+
+locktopus.monk_get = async (db, orig, dmap, slot) => {
+    let meta, data
+    const stmt = db.prepare('SELECT * FROM locks WHERE slot = ?')
+    const lock = stmt.get(slot)
+    lots.push([slot, lock === undefined])
+    if (lock === undefined) {
+        if (last === undefined) locktopus.init_last();
+        [meta, data] = await orig(dmap, slot)
+    } else {
+        meta = lock.meta
+        data = lock.data
+    }
+    return [meta, data]
+}
+
+locktopus.save = async (trace, path) => {
```

Comment:
this whole function is really dense, try to unravel it a bit so that there are fewer expressions per statement